WebMemo

Masaya-Nishijima opened the pull request.

Masaya-Nishijima committed:
Gemfileを管理下に追加

Masaya-Nishijima committed:
既存のメモを表示する機能を記述

Masaya-Nishijima committed:
メモ一覧機能を記述

Masaya-Nishijima committed:
メモの作成機能を記述

Masaya-Nishijima committed:
メモの削除機能を記述

Masaya-Nishijima committed:
メモを表示している画面から削除できるように変更

Masaya-Nishijima committed:
メモの編集機能を記述

Masaya-Nishijima committed:
リファクタリング

コメントの追記
不要機能の削除

Masaya-Nishijima committed:
初回起動用のMakefileを作成

Masaya-Nishijima committed:
デザインを修正

UIを最低限までブラッシュアップ

Masaya-Nishijima committed:
存在しないメモを参照したとき404になるように記述

Masaya-Nishijima committed:
XSS対策を記述

Masaya-Nishijima committed:
READMEを記述

maedana commented on Makefile:
```diff
@@ -0,0 +1,17 @@
+
+$(PUBLIC_DIR)	:
+	mkdir $(PUBLIC_DIR)
+	cp $(STYLESHEET) $(PUBLIC_DIR)/stylesheet.css
```
makeがだめとは言いませんが、rubyなのでRakefileにしてrake使いたいかなーとは思いました。(無理に修正しなくていいです。本題と関係ないですし)

Masaya-Nishijima replied:
Rakefileに移行しました。
(処理自体はUnixコマンドを使用してる箇所が残っています。
ex. `sh %{cp stylesheet.css public/stylesheet.css}`)

Masaya-Nishijima replied:
[cb1f335](https://github.com/Masaya-Nishijima/Sinatra_WebMEMO/pull/1/commits/cb1f3356b30bf4eec00e8f3f0f40a8790d4cdfe4)

maedana commented on README.md:
```diff
@@ -0,0 +1,42 @@
+  - メモの編集
+
+- DELETE `/{メモの名前}`
+  - メモの削除メソッド
```
> ？RESTを意識してURIを作成したつもりですが、私のRESTの理解が正しいか不明です。- GithubのREADMEに今回作成したもののURI構造は記載しています。

と書かれている通りで、全体的にRESTの理解が怪しそうです。RESTの課題もまだ合格してませんし、RESTの課題合格後に全体的に見直してみましょう。

Masaya-Nishijima replied:
RESTを意識して
`/{リソース}/{リソースID}`
の基本形から
`/memo/{memo-name}`
という形に全体を変更しています

Masaya-Nishijima replied:
[714eaa4](https://github.com/Masaya-Nishijima/Sinatra_WebMEMO/pull/1/commits/714eaa43a345e82e195ead5c2b78bb2c68073ad0)

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,163 @@
+require 'sinatra'
+require 'sinatra/reloader'
+
+enable :method_override # メソッドオーバーライドを許可する。
```
クラシックスタイルですし、デフォルトで有効かと思いますよ。
http://sinatrarb.com/intro-ja.html

![image](https://user-images.githubusercontent.com/14562/163889838-b28e2e7d-4aea-4059-bf3d-6d5ed19aee13.png)

Masaya-Nishijima replied:
確認しました。
確かに `enable method_override`を指定しなくても動作致しました。

最初にメソッドオーバーライドがうまく動作しなかったときに、一応設定してそのままになっていたようです。

Masaya-Nishijima replied:
[5434913](https://github.com/Masaya-Nishijima/Sinatra_WebMEMO/pull/1/commits/543491321423b23fa53e6b85f132627d02b0b5d8)

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,163 @@
+get '/not_found' do
+  status 404
+  '404 Not Found'
+end
```
`not_found`ハンドラと`halt`をうまく使いたいかなーと思いました。
http://sinatrarb.com/intro-ja.html
の`エラーハンドリング`の項が参考です。

Masaya-Nishijima replied:
not_foundハンドラとhaltを使用するように変更致しました。

Masaya-Nishijima replied:
[2de2be2](https://github.com/Masaya-Nishijima/Sinatra_WebMEMO/pull/1/commits/2de2be272b56cedbfcb2cd749d4758a21cf0f09a)

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,163 @@
+  <br /> <br/>
+  <a href=\"/new_memo\">メモを作成する<br />
+  "
+  # redirect to('/index.html')
```
ここに限らずですが、`erb`テンプレート使いましょう

Masaya-Nishijima replied:
erbテンプレートを使用するように変更致しました。

SinatraというかWebアプリが初見で、`erbテンプレート`を認知せず、
```ruby
get '/' do
  'Hello world!'
end
```
というベタ書きから延々と拡張しておりました。

Masaya-Nishijima replied:
[0986266](https://github.com/Masaya-Nishijima/Sinatra_WebMEMO/pull/1/commits/098626624cc53f41a19f041a35230f0387eeaf68)

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,163 @@
+  file = File.new("memo_data/#{params[:memo_name]}", 'w')
+  file.write(params[:memo_body])
+  file.close
+  redirect to('/create_memo')
```
メモの作成は作成ページ、削除/更新は一覧ページにそれぞれリダイレクトしており一貫性がないです。こうしたいなら削除は削除完了ページ、更新は完了ページにリダイレクトにするか、作成も一覧画面へのリダイレクトでいいかと。いかがでしょうか?

Masaya-Nishijima replied:
一覧画面へのリダイレクトで統一しました。

CRUDを意識して開発できておらず、
作成→「一般サイトだと作成完了画面ある気がする。」
削除/更新→「まぁ、一覧に戻して、確認したければ見るやろ」
という今思えばダブルスタンダードな考えをしていました。

Sinatraが具体的に、なにができて、なにができないかが理解しておらず、
初めに画面遷移を考えず
メモの表示→メモ一覧→作成→編集→削除と順番に設計していたせいで、
一貫性がなくなったようです。(開発が日を跨いだことも要因)

Masaya-Nishijima replied:
[0a79d7c](https://github.com/Masaya-Nishijima/Sinatra_WebMEMO/pull/1/commits/0a79d7c628d6ab0c7cd6c6d349c52773acea7ae4)

maedana commented:
### その他大きめの設計観点の話

- メモの名前をそのままファイルパスにしてますが、今の仕様では同名メモを保存出来ないですよね。出来れば同名メモが保存できる仕組みを目指してみて欲しいですが難しいでしょうか?

上記を改善すればXSSで言及のある、
> メモの名前がファイルの名前になっているのでそもそも<や>を入力するとSinatraがエラーを吐きます。

も解消するのではないかなと思います。

- 全体的にメモの読み書き部分に`ディレクトリトラバーサル`, `パストラバーサル`に対して脆弱なコードが書かれているように見えます。paramsのような外部から指定可能な値を使って何のチェックもせずにファイルを参照しにいくようなコードは一般的には危険かなと思います。いかがでしょうか? 

参考 https://hawksnowlog.blogspot.com/2020/09/confirm-sinatra-rack-protection-with-path-traversal.html#rackprotectionpathtraversal-の効果

Masaya-Nishijima committed:
MakefileからRakefileへ移行

Masaya-Nishijima committed:
不要な設定を削除

`:method_override`はクラシックスタイルではデフォルトでenable

Masaya-Nishijima committed:
erbファイルを使用するように変更

長々とmyapp.rbに記述していた各HTMLを、erbファイルへ分けて記述するようにした。

Masaya-Nishijima committed:
作成、更新、削除後のリダイレクト先を統一

作成のみ作成完了画面があったが、
作成、更新、削除後は一貫してルートに飛ぶように変更

Masaya-Nishijima committed:
404エラーの出し方を変更

単純なリダイレクトではなくnot_foundハンドラとhaltを利用するように変更

Masaya-Nishijima committed:
RESTfulなURIに変更

/{memo-name}をメモを参照する基本のURIにしていたが
/memo/{memo-name}に変更

Masaya-Nishijima committed:
READMEを修正

修正したコードに合わせてREADMEを変更

Masaya-Nishijima committed:
ファイル操作回りを修正

ディレクトリトラバーサル, パストラバーサル対策として、
`params[:memo_name].delete!("/.<>")`
を追加

既存のメモと同名のメモを作成または編集で発生したとき連番でナンバリングする処理を追加

Masaya-Nishijima commented:
下記の件修正しました(Replyの付け方が理解らなかったので新規コメントです)
- `params[:memo_name].delete!("/.<>")`等で、入力されたメモ名(ファイル名)にディレクトリ関係の`/` `.`やHTMLタグ`<` `>`を削除しその後ファイル操作をするようにした。
- 同名のメモを保存したときに、エラーページに飛ばすのではなく、`{memo_name}-{連番}`で保存するように処理を追加しました。
``` ruby
def generate_unique_name(memo_name)
  return memo_name if !File.exist?("memo_data/#{memo_name}")
  i = 2
  while File.exist?("memo_data/#{memo_name}-" + i.to_s)
    i += 1
  end
  unique_memo_name = memo_name+ '-' + i.to_s
  return unique_memo_name
end
```

> ### その他大きめの設計観点の話
> * メモの名前をそのままファイルパスにしてますが、今の仕様では同名メモを保存出来ないですよね。出来れば同名メモが保存できる仕組みを目指してみて欲しいですが難しいでしょうか?
> 
> 上記を改善すればXSSで言及のある、
> 
> > メモの名前がファイルの名前になっているのでそもそも<や>を入力するとSinatraがエラーを吐きます。
> 
> も解消するのではないかなと思います。
> 
> * 全体的にメモの読み書き部分に`ディレクトリトラバーサル`, `パストラバーサル`に対して脆弱なコードが書かれているように見えます。paramsのような外部から指定可能な値を使って何のチェックもせずにファイルを参照しにいくようなコードは一般的には危険かなと思います。いかがでしょうか?
> 
> 参考 [https://hawksnowlog.blogspot.com/2020/09/confirm-sinatra-rack-protection-with-path-traversal.html#rackprotectionpathtraversal-の効果](https://hawksnowlog.blogspot.com/2020/09/confirm-sinatra-rack-protection-with-path-traversal.html#rackprotectionpathtraversal-%E3%81%AE%E5%8A%B9%E6%9E%9C)

Masaya-Nishijima committed:
リファクタリング

Rubocopの指摘修正

maedana commented on README.md:
```diff
@@ -0,0 +1,42 @@
+  - メモを作成
+
+- ~GET `/create_memo`~
+  - ~メモを正しく作成できたときのリダイレクト先~
```
一覧へのリダイレクトで統一されたということですし、この記述は不要みたいですね

Masaya-Nishijima replied:
確かに不要ですね。
Git&Githubでバージョン管理する習慣がいままでなかったので、
思考錯誤したものは、
ここのようにMarkdownなら打ち消し線を入れたり、
コードならコメントアウトして、
保存しようとする癖が付いていました。

maedana replied:
なるほど気づいてませんでしたがよく見ると打ち消し線だったんですね。Git/Githubなら原則必要なものだけ入れるという考え方で大丈夫かなと思います :+1:

maedana commented on README.md:
```diff
@@ -0,0 +1,42 @@
+  - ~メモを正しく作成できたときのリダイレクト先~
+
+- ~GET `/already_memo`~
+  - ~作成したメモに同名のものがある、空欄のときのリダイレクト先~
```
こちら試行錯誤の過程で使われたものかと思いました。コードを見る限り現在は未使用では?

Masaya-Nishijima replied:
削除致しました。

maedana commented on Rakefile:
```diff
@@ -0,0 +1,31 @@
+task :first_task do
+  puts 'Hello Rakefile!'
+end
```
自動生成のテンプレートっぽいですね。消してよさそうですね

Masaya-Nishijima replied:
削除致しました。

*補足*
自動生成のテンプレートではなく、
こちらも試行錯誤の跡です。

maedana commented on Rakefile:
```diff
@@ -0,0 +1,31 @@
+  sh %{rm Gemfile.lock}
+end
+
+
```
細かいですが空行余計かなと思いますので消しましょうー

Masaya-Nishijima replied:
削除したつもりです。

VS codeで開いたら自動的に消えていました。
(ファイル末尾自動改行の設定のせい?
ならなんで改行ありでコミットされていたのだろう?)

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,107 @@
+# get '/' do
+#   'This is WEB MEMO app'
+#   '<h1>aaaa</h1>'
+# end
```
不要なら消してもよさそうですね。もしくはメモ一覧へのリダイレクトがいいんじゃないですかね

Masaya-Nishijima replied:
メモ一覧へのリダイレクト良い案ですね。

maedana commented on README.md:
```diff
@@ -0,0 +1,42 @@
+
+# 2. URIの構成
+- GET`/memo`
+  - メモ一覧を表示
```
一貫性が有りますし、REST課題も単数で統一しているので修正は任意で結構ですが、僕だったらRailsやその他各種影響により、原則複数形にしてますね。今回だったらURL中の`memo`は原則`memos`にしそうです。単数にするとしたらホントに一個しか存在しえないリソースのときですね。

```
メモ全体 /memos
メモ一件 /memos/:id
```
みたいな。以上ご参考までに

Masaya-Nishijima replied:
やはり、Rails(Ruby)的には、複数要素を持つ物(≒配列的な物)は複数形なのですね。
REST課題と統一している現状を維持しようと思います。

Rubyの感性に慣れていないので、単数・複数形をURI作成のときには意識していませんでした。

個人的な感想ですが、
memoを複数形で見たことがなく、
Linuxのディストリビューション漁りをしていた経験から`/memos`だと、
memというOSのページに見えてしまいます。
> ex. [centos.org](https://www.centos.org/)

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,107 @@
+
+get '/memo' do # メモ一覧の表示
+  erb :memo_list
+  # redirect to('/index.html')
```
```
# redirect to('/index.html')
```
はいらなそうなので削りましょうー

Masaya-Nishijima replied:
削除致しました。

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,107 @@
+end
+
+post '/memo' do # メモを作成
+  params[:memo_name].delete!("/.<>")
```
危険そうなものを手動で削る、という方針にしたんですね。だとすると同じ処理を複数ヶ所でやってますが、もし他にも削るものが発生したときに一箇所の修正になるようにメソッド化するなり、削るものを定数化するなりなどでまとめておきたいですねー

Masaya-Nishijima replied:
定数化で対応致しました。

```ruby
DANGEROUS_STRING = '/.<>'
```

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,107 @@
+end
+
+get '/memo/:memo_name' do # メモを表示
+  redirect(not_found) if read_memo(params['memo_name']) == '指定されたメモがありません'
```
ファイル存在チェックだけをするメソッドが別途あったほうがいいかと思いました。

```ruby
redirect(not_found) if exists_memo?(params['memo_name'])
```
みたいにしたいかなと

Masaya-Nishijima replied:
メソッド`read_memo`後にファイルチェックメソッドを実装致しました。
```ruby
def exists_memo?(memo_name)
  read_memo(params['memo_name']) == '指定されたメモがありません'
end
```

maedana replied:
説明不足で意図が伝わりきってなかったんですが、`指定されたメモがありません`という文字列で判定するのはやめたほうがいいかなということでした。日本語の特定の文字列で判定、というのはちょっとしたことで容易に壊れてしまいそうなロジックかなと。

Masaya-Nishijima replied:
確かにこれでは
下記のメソッドでの文字列に依存していて、文字列が変更されたら動かなくなってしまいますね。
```ruby
def read_memo(memo_name)
  return '指定されたメモがありません' unless File.exist?("memo_data/#{params[:memo_name]}")

  h(File.open("memo_data/#{memo_name}", 'r', &:read))
end
```

Masaya-Nishijima replied:
こちらの処理に変更致しました。
```ruby
def exists_memo?(memo_name)
  !File.exist?("memo_data/#{memo_name}")
end
```

maedana commented on views/memo.erb:
```diff
@@ -0,0 +1,12 @@
+  <p><%= read_memo(params['memo_name']) %></p>
+  </section>
+  <a href="/memo/<%= params['memo_name'] %>/editor">メモを編集する。</a><br />
+  <%= make_delete_form(params['memo_name']) %>
```
こんなことをしなくても`views/delete_form.erb`の内容をここに直接書けばいいだけなのではと思いました。他で使いまわしたりとかがあるわけでもない様子ですし。

Masaya-Nishijima replied:
修正致しました。
htmlをmyapp.rbへ直接書いていた頃に、フォームだけはメソッドに分離していた名残りでした。

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,107 @@
+end
+
+def generate_unique_name(memo_name)
+  return memo_name if !File.exist?("memo_data/#{memo_name}")
```
`if !xxx`は`unless xxx`とかけそうですねー

Masaya-Nishijima replied:
すみませんRubocopが指摘されて修正したの箇所でいした。

Rubocopを通した後に
git pushを忘れていたようです。

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,107 @@
+    i += 1
+  end
+  unique_memo_name = memo_name+ '-' + i.to_s
+  return unique_memo_name
```
最後の`return`は省略可能ですし、Rubyでは省略する文化かなーと思います

Masaya-Nishijima replied:
こちらも同様にRubocopに指摘され修正した箇所です。
git pushを忘れていた箇所

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,107 @@
+  i = 2
+  while File.exist?("memo_data/#{memo_name}-" + i.to_s)
+    i += 1
+  end
```
極端なことを言うと、同じ内容のメモが1万件あったとしたら1万ファイルチェックすることになって凄く遅くなりそうです。こういうケースでも処理が遅くならないようなロジックを考えて見て欲しいかなと思いましたがいかがでしょう?

Masaya-Nishijima replied:
考えましたが思い付いたロジックがどれも一長一短で、自分では現状を変えるほどの物がありませんでした。

下に考えたロジックを載せていますので、思考のヒントを頂けないでしょうか?

考えたロジック
1. 連番ではなく乱数やタイムスタンプをファイル名に付与する。
   - メリット
     - ループが減る。(乱数がダブったときに再試行するだけ)
   - デメリット
     - 乱数なら何桁にする?問題
       - 桁が十分にないと枯渇する。
       - また、枯渇しなくてもナンバリングが進むとダブりが多発する。
     - メモ名に乱数やタイムスタンプが載っているのは美しくない

2. ファイル名=メモ名管理ではなく、ファイル名=IDにしてメモ一つ一つにIDを付与する。
   - メリット
     - 完全同一名のメモができる
       - `めも, めも-2, メモ-3, ...`だけでなく`めも, めも, めも`のように完全一致のメモ名が使える
   - デメリット
     - 完全同一名のメモが、1万件並べられても利点はない。むしろ、メモ名で目的内容のメモを探せないのは、意味がない。
     - ファイルの読み書きすべて修正で実装コストが重い。
     - IDをユニークに生成するのに、被りがないか1万件のファイルを参照することに代りない。(結局、この方法では、さらにロジックを考えないといけない)
 
3. 各メモ名が何番まで使われているかインデックスを作る。(ex. `max("メモ名")`で直近に作成された`メモ名`連番を取れる)
   - メリット
     - 即座に連番を作成できる。
       - 連番のメモ1万件だけしかないなら、インデックスを一行読むだけで作れる
   - デメリット
     - 同名のメモという半ば無駄な存在のために、インデックスを維持するという処理コストは重いのでは?
       - ユニークな(連番されていない)メモが1万件あったらインデックスは1万行
     - 例えばrmコマンドでファイルを削除した場合に、インデックスの整合性が無くなる。

4. すべてのメモを、単一フォルダに纏めて保存する。
   - メリット
     - ファイルチェックは一度だけ
   - デメリット
     - 1万件のメモ名だけでなく、1万件のメモの内容を持った文字列を操作することになる。
     - メモごとにファイルが独立していたほうが、他のシステムとの互換性が良い(Unixならls | grepで特定のファイル名を出し一括操作など)

考慮したこと
- こちらのメモ一覧もメモの総数分のファイルチェック&文字列操作でHTMLタグタグ生成している。
```Ruby
# メモ一覧(ハイパーテキスト)を作成する。
def make_memo_list
  memo_name = Dir.glob('*', base: "#{Dir.getwd}/memo_data")
  list = ''
  memo_name.size.times do |time|
    list += make_a_tag(memo_name[time])
  end
  list
end
# メモの名前からhtmlのaタグリンクを作成するメソッド
def make_a_tag(memo_name)
  "<a href=\"/memo/#{memo_name}\">#{memo_name}</a><br />"
end
```

maedana replied:
思っていた以上に色々検討して頂けたんですね。ありがとうございます :smile: 考えてもらうことが大事だったので今回は実装まではしなくてOKです。もしこれまであまり意識してきてなかったならば、今後は日頃からパフォーマンス面のことも頭の片隅にいれながら実装してもいいかもしれません。とはいえ大抵の場合は実際にパフォーマンスが問題になってから対処すれば十分なケースが多いですし、それだけ利用者が多いってことでもあるなら嬉しい悲鳴ですね。

以下余談です。

1.の場合の乱数は https://docs.ruby-lang.org/ja/latest/method/SecureRandom/s/uuid.html というのを使うと重複する可能性が限りなく低い乱数が使えますよー。`UUID`について調べてみてもいいかもですね。

4.は単一フォルダというか単一ファイルにする実装はアリかなと思います。というのもこうすることによってファイル名を固定出来ますし、ディレクトリ/パストラバーサル脆弱性も考えてなくてよくなるんですよね。grepでの検索性を気にされてますけど、例えば単なるテキストファイルではなく、ymlやjsonなど構造化されたファイルにすれば各種パーサーライブラリありますし、簡単に検索したりなど出来るかなと思います。
ただのテキストファイルだとしても`sed`など使えば正規表現等組み合わせて簡単に置換操作なども出来るかなと思いました。むしろ単一ファイルの方がやりやすいのでは?と個人的には感じます。

maedana commented on views/editor_memo.erb:
```diff
@@ -0,0 +1,15 @@
+</section>
+</article>
+</form>
+<br />
```
ここに限らずですが `<br />`を多用しているのが気になります。改行であれば適切なブロック要素で囲めばいいですし、余白調整のためならcssでmargin, paddingなど活用して調整したいかなと思いました。

Masaya-Nishijima replied:
`<br>`で無理矢理改行していた箇所を
`<div>`や`<h2>`などでマークアップしました。

maedana commented on views/memo.erb:
```diff
@@ -0,0 +1,12 @@
+  "
```
謎のダブルクォーテーションですね。このファイルの末尾にもありますね。

Masaya-Nishijima replied:
削除致しました。

`myApp.rb`に直接書いていた状態から`erb`ファイルへ移行したときの消し忘れです。

maedana commented:
erbについてですが、`<html><body></body></html>`のような基本構造がなく、断片しかないように見えます。ちゃんとしたhtmlとして作成しましょう。なお、共通部分は`layout`ファイルを作るとよいです。

参考
http://sinatrarb.com/intro-ja.html

Masaya-Nishijima committed:
不要な記述を削除

Masaya-Nishijima committed:
リファクタリング

4/28のレビューで、比較的少量の修正となる箇所を纏めて修正

Masaya-Nishijima commented:
> erbについてですが、`<html><body></body></html>`のような基本構造がなく、断片しかないように見えます。ちゃんとしたhtmlとして作成しましょう。なお、共通部分は`layout`ファイルを作るとよいです。
> 
> 参考 http://sinatrarb.com/intro-ja.html

見様見真似で、下記の`layout.erb`を作成しました。
```erb
<!DOCTYPE html>
<html lang="ja">

<head>
  <meta charset="UTF-8">
  <title>WebMemo</title>
  <link rel="stylesheet" href="/stylesheet.css">
</head>

<body>
  <%= yield %>
</body>

</html>
```

Masaya-Nishijima committed:
layout.erbを作成

Masaya-Nishijima committed:
スタイルシートの読み込みをlayout.erbへ移行

Masaya-Nishijima committed:
分離されていたフォームをmemo.erbへ統合

Masaya-Nishijima committed:
htmlを修正

<br />を多用している構成から`div`や`h2`などを利用した構成に変更

Masaya-Nishijima committed:
スタイルシート編集

h2にマージンを設定

Masaya-Nishijima committed:
リファクタリング

rubocop修正

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,106 @@
+def make_delete_form(_memo_name)
+  @memo_name = params['memo_name']
+  erb :delete_form
+end
```
不要になったメソッドは消しましょうー

Masaya-Nishijima replied:
削除致しました。

コーディングから修正まで時間が空いており、全体の構成をあまり意識せず、メソッドが不要になったことに気付けませんでした。
1. メソッド使用箇所を消したとき何も考えずに消した。
2. そもそもメソッドがどこで(何箇所で)利用されているか把握していない。

maedana replied:
この辺は日頃から意識するしかないですねー。型付言語(Java等)だとIDEのサポートによって未使用変数、未使用メソッドなど教えてくれるとは思いますが、Rubyのような動的言語だとそこまでのサポートはないかもですね。(普段IDEは使ってないので詳しくはないですが)

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,106 @@
+  h(File.open("memo_data/#{memo_name}", 'r', &:read))
+end
+
+def exists_memo?(_memo_name)
```
引数渡しているのに利用してないですねー

Masaya-Nishijima replied:
他の指摘と合せて利用するように変更しています。
```ruby
def exists_memo?(memo_name)
  !File.exist?("memo_data/#{memo_name}")
end
```

同様に引数を利用できていない箇所がありましたので修正しています。
```diff
def read_memo(memo_name)
-  return '指定されたメモがありません' unless File.exist?("memo_data/#{params[:memo_name]}")
+  return '指定されたメモがありません' unless File.exist?("memo_data/#{memo_name}")

  h(File.open("memo_data/#{memo_name}", 'r', &:read))
end
```

maedana commented on myapp.rb:
```diff
@@ -0,0 +1,106 @@
+  return memo_name unless File.exist?("memo_data/#{memo_name}")
+
+  i = 2
+  i += 1 while File.exist?("memo_data/#{memo_name}-" + i.to_s)
```
前回コメントの
https://github.com/Masaya-Nishijima/Sinatra_WebMEMO/pull/1#discussion_r860343996
についてはいかがでしょうか?

Masaya-Nishijima replied:
前回修正時、修正コストが大きく後回しそのまま忘れておりました。すみません。
ただいま、[上コメント](https://github.com/Masaya-Nishijima/Sinatra_WebMEMO/pull/1#discussion_r860343996)へ返信しました。

Masaya-Nishijima committed:
リファクタリング
